sdk/helper/kdf/params.go

```diff
@@ -0,0 +1,210 @@
+// previous PRF invocation type (as used in Feedback and Double Pipeline
+// modes).
+//
+// In PKCS#11 v3.0, the counter is split between two parameter types,
+// CK_SP800_108_ITERATION_VARIABLE, a mandatory field, and
+// CK_SP800_108_COUNTER, an optional field ironically forbidden from Counter
+// mode KBKDFs. In a Counter Mode PRF, ITERATION_VARIABLE takes a
+// CounterVariable type, whereas in Feedback or Double Pipeline Modes,
+// ITERATION_VARIABLE takes a NULL value (to signal the previous PRF output,
+// like ChainingVariable does). CK_SP800_108_COUNTER, when allowed, always
+// takes a CounterVariable.
+//
+// Not strictly following the PKCS#11 v3.0 API allows us to define concrete
+// parameter structures holding actual data, rather than using the
+// type+ptr+length indications that PKCS#11 v3.0 prefers.
+
+// Defines an incrementing integer variable; the required incrementing counter
+// variable in Counter mode and an optional variable in Feedback and Double
+// Pipeline modes.
+type CounterVariable struct {
+	// Whether or not to encode this value in little endian or big endian.
+	LittleEndian bool
+
+	// Bit width of the specified variable; must be large enough to hold
+	// the value to encode; must be divisible by 8 with a maximum value of
+	// 64.
+	Width uint8
+}
+
+var _ KBKDFParameter = &CounterVariable{}
+
+// Defines a placeholder value for the previous PRF output (as used in
+// Feedback and Double Pipeline modes). Not permitted in Counter mode.
+type ChainingVariable struct {
+	// No fields.
+}
+
+var _ KBKDFParameter = &ChainingVariable{}
+
+// The DKMLength field type encodes the length of the Derived Key Material.
+// PKCS#11 v3.0 stipulates that no two keys can be produced from the same
+// PRF output segment. This presents a problem for calculating the derived
+// key material's length: is it the sum of the length of all requested
+// keys or is it the sum of the length of all segments used to derive keys?
+// This constant (provided in the DKMLength parameter) decides this.
+//
+// NIST SP800-108's three KBKDFs all use SumOfKeys method with a single key.
+type DKMLengthMethod int
+
+const (
+	// Sum the exact size of all derived keys.
+	SumOfKeys DKMLengthMethod = iota + 1
+
+	// Sum the length of the PRF segments used.
+	SumOfSegments
+)
+
+// Defines the derived key material length field.
+type DKMLength struct {
+	// Which method for calculating DKM length should be used.
+	Method DKMLengthMethod
+
+	// Whether to encode the result in little endian or big endian.
+	LittleEndian bool
+
+	// Bit width of the specified variable; must be large enough to hold
+	// the value to encode; must be divisible by 8 with a maximum value of
+	// 64.
+	Width uint8
+}
+
+var _ KBKDFParameter = &DKMLength{}
+
+// Defines a fixed byte array value (such as the info or context fields from
+// NIST SP800-108). No additional manipulation are performed on these fields.
+type ByteArray struct {
```

Comment:
You can get away with simply "type ByteArray []byte", and implement the interface directly without the wrapper struct.  I might not call it ByteArray though, Go doesn't have arrays, it has slices (and it's very touchy about it)

Reply:
Sans `ChainingVariable`, the remaining types all mirror the PKCS#11v3.0 spec:  https://docs.oasis-open.org/pkcs11/pkcs11-curr/v3.0/pkcs11-curr-v3.0.html#_Toc30061448 

Would you still update this name? If so, to what?